Maple2.File.Parser/AiParser.cs

```diff
@@ -10,24 +10,265 @@
+                        }
+                        // generates an object[] array of the enums. i couldn't find a way to generate the actual underlying enum type array directly because it kept resolving to object[]
+                        var enumObjectValues = Array.ConvertAll(attrib.Value.Split(",", StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries), value => System.Enum.Parse(nestedType, value));
+                        // forcibly create an enum array and copy values over so that field.SetValue is happy & copies the array over to the member
+                        dynamic enumValues = Array.CreateInstance(field.FieldType.GetElementType(), enumObjectValues.GetLength(0));
+                        for (int i = 0; i < enumObjectValues.Length; ++i) {
+                            enumValues[i] = (dynamic)enumObjectValues[i];
+                        }
+                        field.SetValue(outObject, enumValues);
+                    }
+                    else if (field.FieldType.IsEnum) {
+                        field.SetValue(outObject, System.Enum.Parse(field.FieldType, attrib.Value));
+                    }
+                    else {
+                        throw new InvalidDataException($"unhandled type of {field.FieldType.Name} on value '{attrib.Value}' for {typeof(T).Name} member {field.Name}");
+                    }
+                    break;
+            }
+        }
     }
 
     public IEnumerable<(string aiName, NpcAi Data)> Parse() {
         foreach (PackFileEntry entry in xmlReader.Files.Where(entry => entry.Name.StartsWith("AI/"))) {
             string sanitized = Sanitizer.SanitizeAi(xmlReader.GetString(entry));
-            var root = aiSerializer.Deserialize(XmlReader.Create(new StringReader(sanitized))) as NpcAi;
 
-            Debug.Assert(root != null);
+            XmlDocument document = new XmlDocument();
+            var nodeStack = new List<(int StackIndex, XmlNodeList Nodes, object OutNode)>();
+            NpcAi root = new NpcAi();
+
+            document.LoadXml(sanitized);
+
+            nodeStack.Add((0, document.ChildNodes, root));
+
+            while (nodeStack.Count > 0) {
+                (int StackIndex, XmlNodeList Nodes, object OutNode) = nodeStack.Last();
+
+				int stackIndex = nodeStack.Count - 1;
+
+                if (StackIndex >= Nodes.Count)
+                {
+                    nodeStack.RemoveAt(nodeStack.Count - 1);
+
+                    continue;
+                }
+
+                // child node of stack entry
+                XmlNode? xmlNode = Nodes.Item(StackIndex);
+
+                if (xmlNode is null) {
+                    throw new IndexOutOfRangeException($"stack index {StackIndex} node is null [{Nodes.Count}]");
+                }
+
+				bool skipHeader = nodeStack.Count == 1 && xmlNode.Name == "xml";
+
+                if (skipHeader || xmlNode.Name == "#comment") {
```

Comment:
tbh i kind of want comments too so I can copy them to the python scripts, can you make it part of the results and just skip it if you don't need?

Reply:
Yeah I'll add an entry sub type for it